pymodbus/datastore/store.py

```diff
@@ -174,13 +177,13 @@ def getValues(self, address, count=1):
         start = address - self.address
         return self.values[start : start + count]
 
-    def setValues(self, address, values):
+    def setValues(self, address: int, values: Collection):
```

Comment:
I think list[int] is more correct.

This call and the corresponding getValues are only used internally, and do not change independent of how data is stored in the datastore.

Reply:
Yes I know that the documentation suggest you can use setValues() in the sparseDatablock, this is actually a mistake, because it makes people assume they can change the values. setValues() is used internally by the server, and changing it might have side effects on the server.

Changing the values while the server is running will sooner or later cause raising problems, because there are no protection in setValues() to secure access between threads (or tasks). This is more likely with the sparseDataBlock than with the others because you actually replace an array.

The right way is to provide a new method, which should be valid for all datastores, that allows apps to change value/values. Such a method can provide a very generic interface allowing the flexibility needed and of course a lock() to secure the integrity of the data.